### PR TITLE
Updated comparison algorithm and tests for parsers

### PR DESCRIPTION
**Pull request by TarasOrlovskyi**
1. Into DefaultDiscogsService changed search discogs link algoritm
2. Into RawOfferSorter added new comparing algorithm
3. Into CloneNlParser & DecksParser changed getHighResImageLinkFromDocument() method and changed one if() in getInStockInfoFromDocument() method
4. Added DecksParserITest and DecksParserTest and added html pages in
   order to test DecksParserTest
5. Transfer some methods from CloneNlParserTest to CloneNlParserITest

### CODE REVIEW

File: src/main/java/com/vinylteam/vinyl/service/impl/DefaultDiscogsService.java [hunk 1]
```diff
@@ -105,19 +114,35 @@ public String getDiscogsLink(String artist, String release, String fullName) thr
...
+            String discogsFullName = jsonItem.get("title").toString().toLowerCase();
+            String[] preparedDiscogsFullName = Arrays.stream(discogsFullName.split("[- ()!@#$%^&*_+={}:;\"']")).filter(e -> e.trim().length() > 0).toArray(String[]::new);
+            for (String prepareItem : preparedFullNameForMatching) {
+                if (discogsFullName.contains(prepareItem.toLowerCase())){
+                    currentMatching++;
+                }
+            }
+            if (currentMatching > maxMatching){
+                maxMatching = currentMatching;
+                if (((float) maxMatching)/preparedFullNameForMatching.length*100 >= percentMatching
```
Comment (AlexKievDeveloper) [inline src/main/java/com/vinylteam/vinyl/service/impl/DefaultDiscogsService.java]: 100 - looks like a magic number (number without description). Please create field with it.

File: src/test/java/com/vinylteam/vinyl/util/RawOffersSorterTest.java [hunk 2]
```diff
@@ -188,4 +188,88 @@ void addOffersSortingByVinylRawOffersZeroMatchesTest() {
         assertEquals(expectedOffers, actualOffers);
     }
 
+    @Test
+    void fordel() {
```
Comment (AlexKievDeveloper) [inline src/test/java/com/vinylteam/vinyl/util/RawOffersSorterTest.java]: fordel?

File: src/test/java/com/vinylteam/vinyl/util/impl/DecksParserTest.java [hunk 3]
```diff
@@ -17,45 +20,23 @@
...
-    @DisplayName("Get genre links")
-    void getGenresLinks() {
-        //when
-        HashSet<String> actualGenresLinks = decksParser.getGenresLinks();
-        //then
-        assertNotNull(actualGenresLinks);
-    }
-
-    @Test
+        @Test
```
Comment (AlexKievDeveloper) [inline src/test/java/com/vinylteam/vinyl/util/impl/DecksParserTest.java]: Ctrl+Alt+L

File: src/test/java/com/vinylteam/vinyl/util/impl/DecksParserTest.java [hunk 4]
```diff
@@ -100,21 +81,27 @@ void getReleaseFromDocument() {
     @DisplayName("Get catalog number from document")
     void getCatNumberFromDocument() {
         //when
-        String actualCatalogNumberFromDocument = decksParser.getCatNumberFromDocument(testDocument);
+        String actualCatalogNumberFromDocument = decksParser.getCatNumberFromDocument(offerPageDocument);
         //then
         assertNotNull(actualCatalogNumberFromDocument);
     }
 
+        @Test
```
Comment (AlexKievDeveloper) [inline src/test/java/com/vinylteam/vinyl/util/impl/DecksParserTest.java]: Ctrl+Alt+L

File: src/test/java/com/vinylteam/vinyl/util/RawOffersSorterTest.java (hunk 2 above)
Comment (TarasOrlovskyi) [reply src/test/java/com/vinylteam/vinyl/util/RawOffersSorterTest.java]: oops.. yeah.. it was for del :) i forgot about that

File: src/test/java/com/vinylteam/vinyl/util/RawOffersSorterTest.java [hunk 5]
```diff
@@ -188,4 +188,88 @@ void addOffersSortingByVinylRawOffersZeroMatchesTest() {
...
+        //then
+        assertEquals("artist", parameterForComparison);
+    }
+
+    @Test
+    @DisplayName("Return second word 'release' when String is 'A Release'")
+    void getParametersForComparisonWhenStringContainsTwoWordsWithArticleATest() {
+        //when
+        String parameterForComparison = sorter.getParametersForComparison("A Release");
+
```
Comment (Greengrer) [inline src/test/java/com/vinylteam/vinyl/util/RawOffersSorterTest.java]: empty rows like that are not needed.

**Review by Greengrer — CHANGES_REQUESTED** [review]

File: src/main/java/com/vinylteam/vinyl/util/impl/CloneNlParser.java [hunk 6]
```diff
@@ -213,7 +213,14 @@ Boolean getInStockInfoFromDocument(Element document) {
     }
 
     String getHighResImageLinkFromDocument(Element document) {
-        return document.select(HIGH_RES_IMAGE_LINK_SELECTOR).attr("src");
+        String imageLink = document.select(HIGH_RES_IMAGE_LINK_SELECTOR).attr("src");
+        if (imageLink != null && !Objects.equals(imageLink, "")){
+            log.debug("Got high resolution image link {'highResImageLink':{}}", imageLink);
+        } else {
+            log.warn("Can't find image link, returning default");
+            imageLink = "img/goods/no_image.jpg";
```
Comment (TEdwardJ) [inline src/main/java/com/vinylteam/vinyl/util/impl/CloneNlParser.java:225]: the method tries to obtain the link from the source. Probably the case is possible, when the source does not have image link. But I am not sure whether it is a good idea to put mock image on the loading stage; What if we substitute it on view`s data preparation stage?

File: src/test/java/com/vinylteam/vinyl/util/RawOffersSorterTest.java [hunk 7]
```diff
@@ -188,4 +188,65 @@ void addOffersSortingByVinylRawOffersZeroMatchesTest() {
         assertEquals(expectedOffers, actualOffers);
     }
 
+    @Test
+    @DisplayName("Return empty String when parameter is null")
+    void getParametersForComparisonWhenParameterIsNullTest() {
+        //when
+        String parameterForComparison = sorter.getParametersForComparison(null);
+        //then
+        assertEquals("", parameterForComparison);
```
Comment (TEdwardJ) [inline src/test/java/com/vinylteam/vinyl/util/RawOffersSorterTest.java]: isEmpty()

File: src/main/java/com/vinylteam/vinyl/service/impl/DefaultDiscogsService.java [hunk 8]
```diff
@@ -87,14 +88,22 @@ public String getDiscogsLink(String artist, String release, String fullName) thr
...
-        log.debug("Prepared releaseComparing for comparison with artist from Discogs {'releaseComparing':{}}", releaseComparing);
-        query = fullName.replace(" ", "+");
+        artist = artist.toLowerCase();
+        release = release.toLowerCase();
+        fullName = fullName.toLowerCase();
+        String[] preparedFullNameForMatching;
+        if (artist.toLowerCase().contains("various")) {
+            preparedFullNameForMatching = Arrays.stream(release.split("[- ()!@#$%^&*_+={}:;\"']")).filter(e -> e.trim().length() > 0).toArray(String[]::new);
+        } else {
+            preparedFullNameForMatching = Arrays.stream(fullName.split("[- ()!@#$%^&*_+={}:;\"']")).filter(e -> e.trim().length() > 0).toArray(String[]::new);
```
Comment (TEdwardJ) [inline src/main/java/com/vinylteam/vinyl/service/impl/DefaultDiscogsService.java]: The same code, starting from .split() and further... can it be declared in one place?

**Review by AlexKievDeveloper — DISMISSED** [review]

**Review by Greengrer — DISMISSED** [review]

**Review by AlexKievDeveloper — APPROVED** [review]

**Review by littlepony99 — APPROVED** [review]

**Pull request merged**